src/python/nxt_python_asgi.c

```diff
@@ -199,7 +199,153 @@ nxt_python_asgi_init(nxt_unit_init_t *init, nxt_python_proto_t *proto)
+        nxt_unit_alert(NULL,
+                       "Python failed to get '%s' from module 'asyncio'",
+                       runner_class);
+        goto fail;
+    }
+
+    runner = PyObject_CallObject(runner_ref, NULL);
+    if (nxt_slow_path(runner == NULL)) {
+        nxt_unit_alert(NULL, "Python failed to call 'asyncio.%s'",
+                       runner_class);
+        goto fail;
+    }
+    ctx_data->runner = runner;
+    loop_ref = PyObject_GetAttrString(runner, event_loop_func);
+    loop = PyObject_CallObject(loop_ref, NULL);
+        if (nxt_slow_path(runner == NULL)) {
+        nxt_unit_alert(NULL, "Python failed to call 'Runner.%s'",
+                       event_loop_func);
+        goto fail;
+    }
+    for (i = 0; i < nxt_nitems(handlers); i++) {
+        obj = PyObject_GetAttrString(loop, handlers[i].key);
+        if (nxt_slow_path(obj == NULL)) {
+            nxt_unit_alert(NULL, "Python failed to get 'loop.%s'",
+                                 handlers[i].key);
+            goto fail;
+        }
+
+        *handlers[i].handler = obj;
+
+        if (nxt_slow_path(PyCallable_Check(obj) == 0)) {
+            nxt_unit_alert(NULL, "'loop.%s' is not a callable object",
+                                 handlers[i].key);
+            goto fail;
+        }
+    }
+
+    obj = PyObject_CallObject(ctx_data->loop_create_future, NULL);
+    if (nxt_slow_path(obj == NULL)) {
+        nxt_unit_alert(NULL, "Python failed to create Future ");
+        nxt_python_print_exception();
+        goto fail;
+    }
+
+    ctx_data->quit_future = obj;
+
+    obj = PyObject_GetAttrString(ctx_data->quit_future, "set_result");
+    if (nxt_slow_path(obj == NULL)) {
+        nxt_unit_alert(NULL, "Python failed to get 'future.set_result'");
+        goto fail;
+    }
+
+    ctx_data->quit_future_set_result = obj;
+
+    if (nxt_slow_path(PyCallable_Check(obj) == 0)) {
+        nxt_unit_alert(NULL, "'future.set_result' is not a callable object");
+        goto fail;
+    }
+
+    Py_DECREF(loop);
+    Py_DECREF(asyncio);
 
+    *pdata = ctx_data;
+
+    return NXT_UNIT_OK;
+
+fail:
+
+    nxt_python_asgi_ctx_data_free(ctx_data);
+
+    Py_XDECREF(loop);
+    Py_XDECREF(asyncio);
+
+    return NXT_UNIT_ERROR;
+}
+
+
+static void
+nxt_python_asgi_ctx_data_free(void *data)
+{
+    nxt_py_asgi_ctx_data_t  *ctx_data;
+    PyObject *close_loop_func;
+
+    ctx_data = data;
+
+    Py_XDECREF(ctx_data->loop_run_until_complete);
+    Py_XDECREF(ctx_data->loop_create_future);
+    Py_XDECREF(ctx_data->loop_create_task);
+    Py_XDECREF(ctx_data->loop_call_soon);
+    Py_XDECREF(ctx_data->loop_add_reader);
+    Py_XDECREF(ctx_data->loop_remove_reader);
+    Py_XDECREF(ctx_data->quit_future);
+    Py_XDECREF(ctx_data->quit_future_set_result);
+    close_loop_func = PyObject_GetAttrString(ctx_data->runner, "close");
+    PyObject_CallObject(close_loop_func, NULL);
+    Py_XDECREF(ctx_data->runner);
+
+    nxt_unit_free(NULL, ctx_data);
+}
+#else
```

Comment:
This code is almost identical to the other part of the `#else`.  I think it makes more sense to just wrap a few lines in preprocessor conditionals, rather than the entire functions.